mobile/v1/redshift/sql-runner/sql/standard/01-base/01-main/06-events-this-run.sql

```diff
@@ -0,0 +1,183 @@
+/*
+   Copyright 2020-2021 Snowplow Analytics Ltd. All rights reserved.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+DROP TABLE IF EXISTS {{.scratch_schema}}.{{.model}}_base_events_this_run_tmp{{.entropy}};
+
+CREATE TABLE {{.scratch_schema}}.{{.model}}_base_events_this_run_tmp{{.entropy}}
+  DISTSTYLE KEY
+  DISTKEY (event_id)
+  SORTKEY (collector_tstamp)
+AS(
+  {{if eq .model "web"}}
+
+  SELECT
+      a.*,
+      DENSE_RANK() OVER (PARTITION BY a.event_id ORDER BY a.collector_tstamp) AS event_id_dedupe_index --dense_rank to catch event_ids with dupe tstamps later
+
+  FROM
+    {{.input_schema}}.events a
+  INNER JOIN
+    {{.scratch_schema}}.{{.model}}_base_sessions_to_include{{.entropy}} b
+  ON a.domain_sessionid = b.session_id
+
+  WHERE
+    a.collector_tstamp >= (SELECT lower_limit FROM {{.scratch_schema}}.{{.model}}_base_run_limits{{.entropy}})
+    AND a.collector_tstamp <= (SELECT upper_limit FROM {{.scratch_schema}}.{{.model}}_base_run_limits{{.entropy}})
+    AND a.platform IN (
+      {{range $i, $platform := .platform_filters}} {{if $i}}, {{end}} '{{$platform}}' {{else}} 'web' {{end}}
+      )
+    {{if .app_id_filters}}
+    -- Filter by app_id. Ignore if not specified. 
+    AND a.app_id IN (
+      {{range $i, $app_id := .app_id_filters}} {{if $i}}, {{end}} '{{$app_id}}' {{end}}
+      )
+    {{end}}
+  
+  {{end}}
+
+  {{if eq .model "mobile"}}
+  -- Could split mobile_session_ids out into a scratch table.
```

Comment:
My gut says it probably is better to split it. But also...

If we make a scratch table with the mobile sessions data, then we deduplicate that, we don't need to run another window function on the events data.

We would then join that with the events stuff _after_ deduplication.

I realise this is a deviation from having a unified structure across web and mobile - I'll mention that in a top-level comment.

Reply:
All fair points and anything to remove another window function is a good with me. I personally like staying with the unified approach as much as possible. One solution here could be to create the `base_events_this_run_tmp` table in this step using just `com_snowplowanalytics_snowplow_client_session_1` and `base_sessions_to_include` to get the relevant event_ids and tstamps. As you suggest we can then dedupe and finally join in the events table in the `09-add-contexts` stage. This maintains the same steps in the base module, but has slightly differing logic between web and mobile. Thoughts?

Reply:
I think maintaining the steps is a nice approach but where we deviate in a "non-obvious" way, it is worth calling it out specifically in a comment. It's quite likely people will be reading both models and might miss a subtle difference between two steps that have the same name/concept. I'd suggest giving this strategy a go and see how it falls out, it seems like a reasonable suggestion.

Reply:
Yeah I think that's a good approach.

Reply:
I have opened [another PR](https://github.com/snowplow/data-models/pull/57) containing the suggested refactor above.